### ISSUE TITLE
No TeX installation detected when trying to render pdf

### ISSUE THREAD

**Issue by cderv** [quarto-dev/quarto-cli#2408]
### Discussed in https://github.com/quarto-dev/quarto-cli/discussions/2406

<div type='discussions-op-text'>

<sup>Originally posted by **ericvmai** September 12, 2022</sup>
I keep trying to render a quarto html and pdf but rstudio tells me: No TeX installation was detected.

But when I enter quarto install tool tinytex then it says it is already installed. When I try to produce a pdf of a different project - I can produce a PDF with no issue. Help!

https://github.com/hdadvisors/chesterfield-market-analysis/tree/quarto-transition</div>

**Comment by cderv** [comment]
We keep getting some report about similar issues. Can you give us more information : 

* Share `quarto check` output 
* Share `quarto tools info tinytex` output 
* How did you installed TeX on your system ? 
* Which OS are you on ? What is the result of `which tlmgr` if you are on LINUX or `Get-Command tlmgr` in powershell on Windows ?  More generally do you have `tlmgr` found in PATH ? 

Thank you.

**Comment by ericvmai** [comment]
quarto check output:

> [>] Checking basic markdown render....OK
> 
> [>] Checking Python 3 installation....(None)
> 
>       Unable to locate an installed version of Python 3.
>       Install Python 3 from https://www.python.org/downloads/
> 
> [>] Checking R installation...........OK
>       Version: 4.1.2
>       Path: C:/PROGRA~1/R/R-41~1.2
>       LibPaths:
>         - C:/Users/ericv/Documents/R/win-library/4.1
>         - C:/Program Files/R/R-4.1.2/library
>       rmarkdown: 2.16
> 
> [>] Checking Knitr engine render......OK

quart tools info tinytex output:

> {
>   "name": "TinyTeX",
>   "installed": true,
>   "version": "v2022.09",
>   "directory": "C:\\Users\\ericv\\AppData\\Roaming\\TinyTeX",
>   "bin-directory": "C:\\Users\\ericv\\AppData\\Roaming\\TinyTeX\\bin\\win32",
>   "configuration": {
>     "status": "warning",
>     "message": "TeX Live not on path."
>   }
> }

TeX was installed utilizing Yihui's instructions on: https://yihui.org/tinytex/

I'm on Windows OS 11 Pro. Not sure about tlmgr.

**Comment by cderv** [comment]
> "message": "TeX Live not on path."

This is probably the issue. 

Can you open 
* powershell terminal and do `(Get-Command tlmgr).Source`
* or CMD if you prefer and type `where tlmgr`

This is to check if this is found in terminal. I think it will not be found. 

@dragonstyle this seems to be another case where 
* Quarto finds TinyTeX with `quarto tools`
* But it is not on PATH somehow, so it is not found during rendering.

**Comment by ericvmai** [comment]
> > "message": "TeX Live not on path."
> 
> This is probably the issue.
> 
> Can you open
> 
> * powershell terminal and do `(Get-Command tlmgr).Source`
> * or CMD if you prefer and type `where tlmgr`
> 
> This is to check if this is found in terminal. I think it will not be found.
> 
> @dragonstyle this seems to be another case where
> 
> * Quarto finds TinyTeX with `quarto tools`
> * But it is not on PATH somehow, so it is not found during rendering.

> Get-Command : The term 'tlmgr' is not recognized as the name of a cmdlet, function, script file, or operable program.
> Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
> At line:1 char:2
> + (Get-Command tlmgr).Source
> +  ~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : ObjectNotFound: (tlmgr:String) [Get-Command], CommandNotFoundException
>     + FullyQualifiedErrorId : CommandNotFoundException,Microsoft.PowerShell.Commands.GetCommandCommand

**Comment by cderv** [comment]
Yes so the issue is that for some reason, Installation was not complet and `tlmgr` has not be added to PATH. 

This leads to Quarto not finding it. Is this possible to uninstall TinyTeX and reinstall using `quarto install tool tinytex` ? 

Installation should call `tlmgr path add` to correctly put the binary into the PATH. (note that on Windows, you may need to restart Terminal or even computer for this to work after installation)

**Comment by ericvmai** [comment]
What method do you suggest for uninstalling TinyTeX?

**Comment by cderv** [comment]
if you used `tinytex::install_tinytex()`, you can do `tinytex::uninstall_tinytex()`.

**Comment by ericvmai** [comment]
```
compilation failed- error
Extra }, or forgotten \endgroup.                p
\@endfloatbox ...pagefalse \outer@nobreak \egrou  \color@endboxp

  \color@endbox
l.834 \end{table}  
```

[index.log](https://github.com/quarto-dev/quarto-cli/files/9558688/index.log)

**Comment by cderv** [comment]
It seems the issue is solved, and TinyTeX is found for Quarto to render PDF. That is good to know ! 

You now seem to have another type of issue, which are LaTeX issues. This is different. 

Either there could be some versions issues (please to update to latest Quarto if possible). Or it could be some problem without the qmd content. 

Which document are you rendering ?

**Comment by nzgwynn** [comment]
I'm teaching beginner students (their first R class) and they are having the same issue. What is the best way to help them debug?

**Comment by cderv** [comment]
> they are having the same issue

You mean the no TeX installation detected ? 

We are currently working on improving all this in Quarto so that when TinyTeX is found, it will be used. 

Currently, Quarto assumes that TeX Live is available by looking at `tlmgr`  on PATH - So it must be on PATH. 

To debug: 

* Which TeX Live installation are they running ? If none they install, it could be good to use the TinyTeX one `quarto tools install tinytex`
* If still not working after this installation, add on PATH tlmgr - usually `tlmgr path add` is the solution, but it should have been run by the installer. 
* Uninstalling and reinstalling TinyTeX is an option as it seems that solved the issue for @ericvmai above. 
* Adding manually on PATH the bin directory where `tlmgr` is should be possible too. 


So basically, the thing is for now **`tlmgr` should be found on PATH**. For later version of Quarto, it will be improved and hopefully this will never happen again

Hope it helps

**Comment by nzgwynn** [comment]
We've uninstalled and reinstalled a few times using `quarto install tool tinytex` which didn't help.

You're saying in the console run 

```
tinytex::uninstall_tinytex()
tinytex::install_tinytex() 
```
Is the best next step?

**Comment by cderv** [comment]
How was TinyTeX installed in the first place ? Using the **tinytex** R package ? 
If so, then yes you can try this step. `tlmgr path add` should be run by default as part of the installation steps. 

You can still try to run it yourself. For R users managing with R package: 

* `tinytex::tinytex_root()` should return where the installation took place
* `tinytex::is_tinytex()` will confirm that it is a tinytex
* `tinytex::tlmgr_path("add")` should run the command to add in PATH. 

Not that going to `tinytex::tinytex_root()` folder, you can find the `tlmgr` binary and then run the command from terminal there too 
```
tlmgr path add
```

I am sorry this looks so complex. I am trying to give as much information as I can to help without seeing the real environment and trying to see the real issues.

**Comment by ericvmai** [comment]
> 

I've pinpointed my issue down to the part-1a-2.qmd file in the referenced repo. I cannot for the life of me find any extra } or other issues that would be causing a problem.

**Comment by ericvmai** [comment]
> It seems the issue is solved, and TinyTeX is found for Quarto to render PDF. That is good to know !
> 
> You now seem to have another type of issue, which are LaTeX issues. This is different.
> 
> Either there could be some versions issues (please to update to latest Quarto if possible). Or it could be some problem without the qmd content.
> 
> Which document are you rendering ?

The document I'm rendering is part-1a-r.qmd and the index.qmd. I cannot pinpoint an extra bracket anywhere and there aren't any issues on the line referenced as far as I can tell.

**Comment by dragonstyle** [comment]
I'm a little out of my depth re Kable tables, but from looking at the generated TeX, I'm guessing that both Quarto and Kable are attempting to apply captions (and something about that isn't going well):

```markdown

```{r}
#| label: tbl-cost-burden
#| tbl-cap: "Cost burden by tenure in Chesterfield County"

....

chesterfield_cb_table %>% 
  rename('Cost burden' = 1) %>% 
  kable(caption = "Cost burden by tenure in Chesterfield County") %>%                       
  kableExtra::kable_styling(
    bootstrap_options = c("condensed", "hover", "striped")) %>% 
  kableExtra::row_spec(4, bold=TRUE)
```

Here's the generated table caption:

```tex
\caption{\label{tbl-cost-burden}Cost burden by tenure in Chesterfield County }Cost burden by tenure in Chesterfield County}
```
which contains the phantom extra brace...

**Comment by nzgwynn** [comment]
I found a work around @AmeliaMN! Have them render an HTML, you can open the HTML in a website and save it as a pdf! They can open in Chrome and hit print and to save to a pdf!

**Comment by ericvmai** [comment]
> I'm a little out of my depth re Kable tables, but from looking at the generated TeX, I'm guessing that both Quarto and Kable are attempting to apply captions (and something about that isn't going well):
> 
> ```gfm
> ```{r}
> #| label: tbl-cost-burden
> #| tbl-cap: "Cost burden by tenure in Chesterfield County"
> 
> ...
> 
> chesterfield_cb_table %>% 
>   rename('Cost burden' = 1) %>% 
>   kable(caption = "Cost burden by tenure in Chesterfield County") %>%                       
>   kableExtra::kable_styling(
>     bootstrap_options = c("condensed", "hover", "striped")) %>% 
>   kableExtra::row_spec(4, bold=TRUE)
> ```
> ```
> 
> Here's the generated table caption:
> 
> ```
> \caption{\label{tbl-cost-burden}Cost burden by tenure in Chesterfield County }Cost burden by tenure in Chesterfield County}
> ```
> 
> which contains the phantom extra brace...

Thank you! @dragonstyle  this solved my issues and it successfully rendered to pdf!

**Comment by cderv** [comment]
We have added better detection process for the TinyTeX installation, installed by Quarto or by R with **tinytex**

This currently lives in the pre release for 1.2 that can be downloaded at https://quarto.org/docs/download/

Changes are: 
* If a TinyTeX installation is found, it will be used by Quarto even if not in PATH. 
* `quarto tools install tinytex` won't put in PATH by default to avoid problem with other installation. The former point insure TinyTeX will be used. 
* Other changelog at https://github.com/quarto-dev/quarto-cli/blob/main/news/changelog-1.2.md#tinytex

Please do give a try to the pre-release if you still encounter issue with some installation.

Thank you everyone for your feedback. This helped us improve !

**Comment by GShotwell** [comment]
I ran into this issue and it turned out that my /usr/local/bin was not writable. This old tinytex issue resolved it for me. 

https://github.com/rstudio/tinytex/issues/24

**Comment by cderv** [comment]
> Do I need python/pip/jupyter for compiling quarto documents with only R chunks?

No you don't need Python / Jupyter is you don't need document with Python. 

> The error if I hit the "Render project" button in RStudio is:

It seems you have a valid installation of LaTeX - See https://quarto.org/docs/output-formats/pdf-engine.html#installing-tex

> Version: 1.2.335

Please do update to 1.3 Release Candidate available at https://quarto.org/docs/download/ - it should have some fix toward better error message I believe.

**Comment by memeplex** [comment]
> Share `quarto tools info tinytex` output

@cderv  This was an useful tip given that I was struggling to find the location of tinytex. But FYI it's not documented `quarto tools --help`.

**Comment by mcanouil** [comment]
@memeplex are you using the version of Quarto from last year? Be cautious of the date/version.
The posts mentioned 1.2 (stable) and a 1.3 pre-release. We are at 1.4 (stable).

**Comment by memeplex** [comment]
1.4.549. It was just a heads up.

**Comment by cderv** [comment]
> But FYI it's not documented quarto tools --help.

Yes we know. quarto CLI help needs some improvements, especially on some specific commands like this. Thanks for the heads up.
